File: src/medium/Q1209/Q1209.cpp
// 1209. Remove All Adjacent Duplicates in String II
// https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
#include <string>
#include <utility>
#include <vector>

using namespace std;

// Time: O(n), Space: O(n), where n is the size of `s`.
auto remove_duplicates(string s, int k) -> string {
  auto stack = vector<pair<char, size_t>>();
  for (const auto c : s) {
    if (stack.empty()) {
      stack.emplace_back(c, 1);
      continue;
    }
    auto &top = stack.back();
    if (top.first != c) {
      stack.emplace_back(c, 1);
      continue;
    }
    if (top.second + 1 == k) {
      stack.pop_back();
      continue;
    }
    // Increment count of 'c'
    ++top.second;
  }
  string result;
  for (const auto &[c, count] : stack) {
    result += string(count, c);
  }
  return result;
}

File: src/main/python/medium/Q1209.py
"""
1209. Remove All Adjacent Duplicates in String II
https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string-ii/
"""


# Time: O(n), Space: O(n), where n is the size of `s`.
def remove_duplicates(s: str, k: int) -> str:
    # Stack containing (char, count) tuples
    stack: list[tuple[str, int]] = []
    for c in s:
        if len(stack) == 0:
            item = (c, 1)
            stack.append(item)
            continue
        top: tuple[str, int] = stack[-1]
        if top[0] != c:
            item = (c, 1)
            stack.append(item)
            continue
        count = top[1] + 1
        if count == k:
            stack.pop()
            continue
        item = (c, count)
        stack[-1] = item
    result = ""
    for item in stack:
        c, count = item
        result += c * count
    return result
